### DISCUSSION TITLE
label weighted loss함수

### DISCUSSION THREAD

**Discussion by papari1123** [boostcampaitech4nlp2/level1_semantictextsimilarity_nlp-level1-nlp-13#41] [Ideas]
이 discussion은 대회 마감 전에 얻은 인사이트라서,  최근 정리한 링크를 참조하면 좋을 거 같습니다.
[링크](https://ebony-dime-3e7.notion.site/1-ensemble-359df59ed9b3454289f9fd83c4da3bde)

# 요약
- label weighted loss함수로 validation에서 예측못했던 label(2-3)에 대해서 예측 성능을 올릴 수 있었지만, (당연하게도) 다른 label에서의 예측성능이 조금 떨어졌다. 
- label별 예측에 특화된 모델 5개(0-1, 1-2, .., 4-5 label)를 앙상블해서 voting을 하고, voting 결과에 맞는 모델의 예측값을 결과로 활용하면 좋을 듯하다.

# label weighted loss 함수(가칭)에 따른 결과
## [아이디어]
- 기존 실험(roberta/large + MSE)의 결과를 분석했을 때, label 2~3에서의 예측 능력이 떨어졌고 예측 분포 또한 validation label 분포 대비해 imbalence했다. 
- 이를 해결하기 위해 loss 함수에 label에 대한 가중치를 줘서, 학습을 하고 결과를 분석하기로 했다.
- MSE와 MAE의 차이도 같이 확인하여 더 좋은 것을 사용하기로 하였다.
```python
class wMSELoss(nn.Module): # for imbalanced data
    def __init__(self, weight=5, bias=0.1):
        super(wMSELoss, self).__init__()
        self.weight = weight

    def forward(self, inputs, targets):
        mse_func = nn.MSELoss(reduce=False)
        mse_loss = mse_func(inputs, targets)
        mse_loss =  torch.dot(mse_loss, 1/(bias + abs(targets - 2.5)*weight))
        return mse_loss
```
## [실험가설]
- 상대적으로 예측이 어려운 label 2-3 구간에 대해 loss 가중치를 크게 주면, 모델이 모든 label 구간에 대해 좀 더 균일한 예측 결과를 얻을 수 있을 것이다.
 
## [실험결과]
- valid_pearson 그래프
    - weightedMAE는 0.905가 나오고, weightedMSE는 early stopping 없이 50epoch까지 pearson 계수가 업데이트되었고, 0.9305가 나왔다.
![image](https://user-images.githubusercontent.com/33012030/199165696-42466ccc-9835-46c1-b900-63ff484633c9.png)

※  여기서 언급되는 MSE, MAE는 label weighted 되어 계산된 것이라, weighted 되지 않는 MAE, MSE loss와 다른 경향성을 나타낼 수 있다.
- 통계량 (좌: MAE, 우:MSE)
  - MSE가 diff(예측값과 label값 차이)가 낮게 나왔지만, pred(예측값)의 std값이 상대적으로 높기 때문에 피어슨 상관계수를 계산할 때 좀 불리한 면이 있다. 예측값의 std를 줄이기 위해 앙상블 기법을 쓸 수도 있겠다. 
![image](https://user-images.githubusercontent.com/33012030/199161288-8cadbab7-1d1f-4b04-a2bc-ac96fdaa8ed2.png)
- 피어슨 상관계수 (출처 : [위키백과](https://ko.wikipedia.org/wiki/%ED%94%BC%EC%96%B4%EC%8A%A8_%EC%83%81%EA%B4%80_%EA%B3%84%EC%88%98))
![image](https://user-images.githubusercontent.com/33012030/199161529-63a11464-a91a-4e6c-b815-605754efe851.png)

-  scatter plot
  - MSE가 편차를 제곱하기 때문에 outlier에 민감하다고 하더니 실제로 결과가 그렇게 나왔다.
  ![image](https://user-images.githubusercontent.com/33012030/199158487-ee6b7728-7b31-4980-86d2-3e11a8e329f5.png)
- box/violin plot
- MSE의 경우, 학습에 따른 개선 결과를 확인하게 위해, 중간 결과(epoch=10)를 그래프에 추가하였다.
- box plot에도 MSE는 label과 예측값(pred)간의 차이(diff)가 MAE보다 작음을 확인할 수 있다. IQR도 작게 나온다. 
![image](https://user-images.githubusercontent.com/33012030/199158881-c1764bc8-0f98-4f08-ad50-0169e99570d3.png)

- 반면 **예측값에 대한 box plot을 보면 MAE가 IQE가 작다**. 또한 median값이 2.5 부근에 있어서, MAE는 정직하게 loss함수의 weight값을 따라간다.
![image](https://user-images.githubusercontent.com/33012030/199159803-9cb77410-0c35-417a-b921-4ebb864ced8a.png)

- 예측값의 분포를 자세히 확인하기 위해 violin plot을 사용했다. 먼저 label 분포를 확인하면 0~5범위에 대해 validation label이 균등하게 포함된 상황이다.
![image](https://user-images.githubusercontent.com/33012030/199162401-9ef84a9c-87a8-4f21-9108-ea50bb3a9ee7.png)
- 분포를 확인해보면, MSE_last로 갈수록, 모든 label에 대해 균등하게 예측함을 알 수 있다. label에 대해 weighted된 loss함수임에도 불구 하고 MSE는 균등하게 label을 예측하는 이유를 추측하자면,
    - (추가 예정)
![image](https://user-images.githubusercontent.com/33012030/199161998-e87e7124-2815-4821-a218-58a9a25a90e6.png)

- MSE의 label 범위에 따른 diff 분포
![image](https://user-images.githubusercontent.com/33012030/199163376-d8bec13b-4102-4215-9697-6006f025acb4.png)
- MAE의 label 범위에 따른 diff 분포  : MSE보다 2-3 label에서 예측을 잘하나, 다른 label에서는 못한다.
![image](https://user-images.githubusercontent.com/33012030/199163447-01f48ed7-f4f2-4779-96ec-6ab8f638b5b8.png)

※ 위의 결과를 봤을 때 weightMSE의 weight가 무슨 의미가 있냐는 생각을 할 수 있는데, 아래 [label weight없이 MSE로 나온 결과](https://github.com/boostcampaitech4nlp2/level1_semantictextsimilarity_nlp-level1-nlp-13/discussions/27)와 달리, label의 분포처럼 pred의 분포가 균일해졌다. (다만, 다른 label 구간에서의 예측값 차이가 소폭 상승했다. trade-off 느낌)
![image](https://user-images.githubusercontent.com/33012030/199164667-8ac819aa-979a-46a3-9b77-4e4240932b67.png)
![image](https://user-images.githubusercontent.com/33012030/199164650-d6538c8d-e29c-4a60-b3ba-1d1c3279fe13.png)

## [인사이트]

#  SR augmentaion에 따른 결과
## [인사이트]

**Comment by ymnseol** [comment]
Label별 앙상블 좋은 생각입니당!! 우리 모델이 못 맞추는 구간이 있으니까 그 부분 데이터만 학습하는 모델 + 원래 모델 앙상블하면 어떨까하고 생각하고 있었는데,,, 구체화된 인사이트 공유 감사합니다👍 앙상블하려면 모델은 조금 작은 걸 써야하려나,,,?
